qcodes_contrib_drivers/drivers/Vaunix/LDA.py

```diff
@@ -0,0 +1,291 @@
+                 **kwargs):
+        r"""
+        QCoDeS Instrument for Vaunix LDA digital attenuators.
+
+        Args:
+            name: Qcodes name for this instrument
+            serial_number: Serial number of the instrument, used to identify
+                it.
+            dll_path: Look for the LDA DLLs in this directory.
+                By default, the directory this file is in.
+            channel_names: Optionally assign these names to the channels.
+            test_mode: If True, communicates with the API but not with the
+                device. For testing purposes.
+        """
+
+        super().__init__(name=name, **kwargs)
+        self.serial_number = serial_number
+        self.reference = None
+
+        if channel_names is None:
+            channel_names = {}
+
+        self.dll = self._get_dll(dll_path)
+        self.dll.fnLDA_SetTestMode(test_mode)  # Test API without communication
+
+        # Find all Vaunix devices, init the one with matching serial number.
+        num_devices = self.dll.fnLDA_GetNumDevices()
+        device_IDs = ctypes.c_int * num_devices
+        device_refs = device_IDs()
+        self.dll.fnLDA_GetDevInfo(device_refs)
+        devices = {self.dll.fnLDA_GetSerialNumber(ref): ref
+                   for ref in device_refs}
+        self.reference = devices.get(self.serial_number, "not found")
+        if self.reference == "not found":
+            raise ValueError(f"LDA with serial number {self.serial_number}"
+                             f" was not found in the system. Found: {devices}")
+
+        self.dll.fnLDA_InitDevice(self.reference)
+
+        num_channels = self.dll.fnLDA_GetNumChannels(self.reference)
+        if num_channels == 1:
+            # don't add Channel objects, add parameters directly instead
+            self.add_parameter("attenuation",
+                               parameter_class=LdaAttenuation,
+                               )
+            wf_vals = LdaWorkingFrequency.get_validator(self)
+            if wf_vals:
+                self.add_parameter("working_frequency",
+                                   parameter_class=LdaWorkingFrequency,
+                                   vals=wf_vals,
+                                   )
+        else:
+            for i in range(1, num_channels + 1):
+                name = channel_names.get(i, f"ch{i}")
+                ch = LdaChannel(parent=self, channel_number=i, name=name)
+                self.add_submodule(name, ch)
+
+        self.connect_message()
+
+    def _get_dll(self, dll_path: str = None) -> ctypes.CDLL:
+        r"""
+        Load correct DLL from ``dll_path`` based on bitness of the operating
+        system.
+
+        Args:
+            dll_path: path to the directory that contains the Vaunix LDA DLL.
+                By default, same as the directory of this file.
+        """
+        if dll_path is None:
+            dll_path = os.path.dirname(os.path.abspath(__file__))
```

Comment:
a good practice is to expose the dll_path as a class attribute, so that users can set it before initializing the instrument AND don't need to pass it to the instrument's init each time.

Reply:
Fixed in 0191808.